lib/model/compose.dart

```diff
@@ -0,0 +1,179 @@
+
+  // (A) Why not panic on dangling opening fences
+  //     (ones without a corresponding closing fence)?
+  //     Because the spec allows leaving the closing fence implicit:
+  //     > If the end of the containing block (or document) is reached
+  //     > and no closing code fence has been found,
+  //     > the code block contains all of the lines after the opening code fence
+  //     > until the end of the containing block (or document).
+  //
+  // (B) Why not look for dangling closing fences (ones without an opening fence)?
+  //     Because technically there's no such thing:
+  //     they would be indistinguishable from dangling opening fences,
+  //     and parsers (and this function) will treat them that way.
+  final fenceLength = getUnusedOpeningBacktickFenceLength(content);
+
+  resultBuffer.write('`' * fenceLength);
+  if (infoString != null) {
+    resultBuffer.write(infoString);
+  }
+  resultBuffer.write('\n');
+  resultBuffer.write(content);
+  resultBuffer.write('\n');
+  resultBuffer.write('`' * fenceLength);
+  resultBuffer.write('\n');
+  return resultBuffer.toString();
+}
+
+const hashReplacements = {
+  "%": ".",
+  "(": ".28",
+  ")": ".29",
+  ".": ".2E",
+};
+
+// Corresponds to encodeHashComponent in Zulip web;
+// see web/shared/src/internal_url.js.
+String encodeHashComponent(String str) {
+  return Uri.encodeComponent(str)
+    .replaceAllMapped(RegExp(r'[%().]'), (Match m) => hashReplacements[m[0]!]!);
+}
+
+/// A URL to the given [Narrow], on `store`'s realm.
+///
+/// To include /near/{messageId} in the link, pass a non-null [nearMessageId].
+// Why take [nearMessageId] in a param, instead of looking for it in [narrow]?
+//
+// A reasonable question: the "near" part of a near link (e.g., for
+// quote-and-reply) takes the same form as other operator/operand pairs that we
+// represent with [ApiNarrowElement]s, like "/stream/48-mobile".
+//
+// But unlike those other elements, we choose not to give the "near" element
+// an [ApiNarrowElement] representation, because it doesn't have quite that role:
+// it says where to look in a list of messages, but it doesn't filter the list down.
+// In fact, from a brief look at server code, it seems to be *ignored*
+// if you include it in the `narrow` param in get-messages requests.
+// When you want to point the server to a location in a message list, you
+// you do so by passing the `anchor` param.
+Uri narrowLink(PerAccountStore store, Narrow narrow, {int? nearMessageId}) {
+  final apiNarrow = narrow.apiEncode();
+  final resultBuffer = StringBuffer('#narrow');
+  for (final element in apiNarrow) {
+    resultBuffer.write('/');
+    if (element.negated) {
+      resultBuffer.write('-');
+    }
+    // TODO(server-7) remove special-casing
+    if (element is ApiNarrowDm) {
+      final supportsOperatorDm = store.connection.zulipFeatureLevel! >= 177; // TODO(server-7)
+      final resolved = element.resolve(legacy: !supportsOperatorDm);
```

Comment:
I think this logic can actually be made somewhat cleaner by accepting a bit of mutation:

```suggestion

    if (element is ApiNarrowDm) {
      final supportsOperatorDm = store.connection.zulipFeatureLevel! >= 177; // TODO(server-7)
      element = element.resolve(legacy: !supportsOperatorDm);
    }
```

Then the rest of the logic fits into the same structure as the other narrow-element types. The operator gets written the same way as always, and the operand can look like:
```dart
      case ApiNarrowDmModern():
        final suffix = element.operand.length >= 3 ? 'group' : 'dm';
        resultBuffer.write('${element.operand.join(',')}-$suffix');
      case ApiNarrowPmWith():
        final suffix = element.operand.length >= 3 ? 'group' : 'pm';
        resultBuffer.write('${element.operand.join(',')}-$suffix');
      case ApiNarrowDm():
        assert(false, 'ApiNarrowDm should have been resolved');
```